lib/identity_key/identity_key.c

```diff
@@ -0,0 +1,207 @@
+/*
+ * Copyright (c) 2022 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include <string.h>
+#include <identity_key.h>
+#include <nrf_cc3xx_platform_kmu.h>
+#include <nrf_cc3xx_platform_ctr_drbg.h>
+#include <nrf_cc3xx_platform_identity_key.h>
+#include <hw_unique_key.h>
+#include <zephyr/sys/util.h>
+#include <psa/crypto.h>
+
+#ifdef __ZEPHYR__
+#include <zephyr/sys/printk.h>
+#include <zephyr/kernel.h>
+#define IDENTITY_KEY_PRINT(msg) printk(msg)
+#define IDENTITY_KEY_PRINT_VAL(msg, val) printk(msg " %d", val)
+#define IDENTITY_KEY_PANIC k_panic
+
+#else
+static void panic(void)
+{
+	while (1)
+		;
+}
+#define IDENTITY_KEY_PRINT(...)
+#define IDENTITY_KEY_PANIC panic
+
+#endif /* __ZEPHYR__ */
+
+/* Use 'copy_of_uicr_word_read' instead of 'nrfx_nvmc_uicr_word_read'
+ * until it has become available everywhere. Everywhere consists of
+ * TF-M's copy of nrfx, nrfx, hal_nordic in Zephyr and NCS.
+ */
+
+static uint32_t copy_of_uicr_word_read(uint32_t const volatile * address)
+{
+	uint32_t value = *address;
+
+#if NRF91_ERRATA_7_ENABLE_WORKAROUND
+	__DSB();
+#endif
+
+	return value;
+}
+
+#ifdef CONFIG_IDENTITY_KEY_RANDOM
+static int generate_random_secp256r1_private_key(uint8_t *key_buff)
+{
+	psa_status_t status;
+	psa_key_handle_t key_handle;
+	size_t olen;
+
+	/* Initialize PSA Crypto */
+	status = psa_crypto_init();
```

Comment:
This is following statements from the PSA Crypto API specifications

> It is recommended that applications call this function before calling any other function in this module.
> 
> Applications are permitted to call this function more than once. Once a call succeeds, subsequent calls are guaranteed to succeed.

Reply:
I think I was misunderstood, let me try to clarify.

If we were to call this at boot, e.g. using the SYS_INIT system, then any code that was known to run after/in main could omit calling psa_crypto_init because it would already be called.

This would be compatible with the spec. AFAICT.

This is best-practice for NCS libraries/drivers but there are circumstances where it should be avoided, it's not clear to me if there is any reason to not follow the best practice in this case.

NB: Still out-of-scope for this PR.